geps/gep-1742.md

```diff
@@ -301,13 +301,167 @@ Could not find any HTTP specific timeouts. PRs welcomed. 😊
 
 ## API
 
-TBD.
+The above diagrams show that there are many different kinds of configurable timeouts
+supported by Gateway implementations: connect, idle, request, upstream, downstream.
+Although there may be opportunity for the specification of a common API for more of
+them in the future, this GEP will focus on the L7 timeouts in HTTPRoutes that are
+most valuable to clients.
+
+From the above analysis, it appears that, at a minimum, implementations are capable of
+supporting the configuration of simple client downstream request timeouts on HTTPRoute
+rules. This is a relatively small addition that would benefit many users.
+
+A client HTTP request to a gateway may result in more than one call to the destination
+backend service, for example, if automatic retries are supported. Some implementations
+support configuring a timeout for the individual backend requests, seperate from the
+overall client request timeout. Adding non-Core support for this would also benefit
+many users.
+
+### GO
+
+```go
+type HTTPRouteRule struct {
+	// Timeouts defines the timeouts that can be configured for an HTTP request.
+	//
+	// Support: Core
+	//
+	// +optional
+	// <gateway:experimental>
+	Timeouts *HTTPRouteTimeouts `json:"timeouts,omitempty"`
+
+	// ...
+}
+
+// HTTPRouteTimeouts defines timeouts that can be configured for an HTTPRoute.
+// Timeout values are formatted like 1h/1m/1s/1ms as parsed by Golang time.ParseDuration
+// and MUST BE >= 1ms.
+type HTTPRouteTimeouts struct {
+	// Request specifies a timeout for the Gateway to send a response to a client HTTP request.
+	// Whether the gateway starts the timeout before or after the entire client request stream
+	// has been received, is implementation dependent.
+	//
+	// For example, setting the `rules.timeouts.request` field to the value `10s` in an
+	// `HTTPRoute` will cause a timeout if a client request is taking longer than 10 seconds
+	// to complete.
+	//
+	// Request timeouts are disabled by default.
```

Comment:
This may be problematic. I think many implementations likely already set default timeouts (for example GKE Gateway uses 30 seconds by default). 

Unfortunately we can't set a default on a new field in a backwards compatible way as that would change the behavior of the API during an upgrade.

I think we may need to loosen this up to say something like "When this field is unspecified, request timeout behavior is implementation-specific."

Would be helpful to see if other implementations are already setting defaults here.

/cc @arkodg @howardjohn @sunjayBhatia @mlavacca @shaneutt @youngnick

Reply:
As far as I remember Envoy doesn't have an exact analogy for this but we might approximate to use something like the "response timeout" field which measures the time taken for the backend to return a response and has a default of 15s: https://projectcontour.io/docs/1.25/config/api/#projectcontour.io/v1.TimeoutPolicy

Reply:
> there will be ways to override the default (e.g., a policy attachment on the Gateway

Note that policy attachment GEP intentionally avoids configuring anything that overlaps with config in Gateway API resources. In its current form, it's meant to extend Gateway API, not override existing values. If we ever extend the scope of policy attachment to include this, it will only be after we can make it _very easy_ for users to understand what's happening, primarily to avoid issues like @kflynn described in #2015.

Reply:
Understood, but I would argue that the default values do not overlap with config in Gateway API resources unless we also provide some API to set them. Regardless, I think your point is correct that we unfortunately need to say that the default is implementation-specific.

Reply:
@robscott changed default to implementation-dependent.